### DISCUSSION TITLE
Add REST client headers as attributes to OpenTelemetry span for REST call

### DISCUSSION THREAD

**Discussion by svifred** [quarkusio/quarkus#44256] [Q&A] [answered]
I would like to add attributes, specifically client request and response headers, to the OpenTelemetry span for each REST call. For the OpenTelemetry Java agent instrumentation, I would take advantage of the [configuration parameters](https://opentelemetry.io/docs/zero-code/java/agent/instrumentation/http/)

```
otel.instrumentation.http.client.capture-request-headers
otel.instrumentation.http.client.capture-response-headers
```

but I suspect the Quarkus instrumentation does not provide an implementation of these, as an application.properties with

```
quarkus.otel.instrumentation.http.client.capture-request-headers=x-request-header
quarkus.otel.instrumentation.http.client.capture-response-headers=x-response-header
```

results in

```
WARN Unrecognized configuration key "quarkus.otel.instrumentation.http.client.capture-response-headers" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
WARN Unrecognized configuration key "quarkus.otel.instrumentation.http.client.capture-request-headers" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
```

I have tried implementing a `ClientResponseFilter` utilizing `Span.current()`, but attributes set on this span is found _not_ on the span for the REST client call, but on the _parent_ span. When adapting the implementation to leverage the `Context` added as a property on the request by the Quarkus instrumentation [OpenTelemetryClientFilter](https://github.com/quarkusio/quarkus/blob/main/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/intrumentation/restclient/OpenTelemetryClientFilter.java#L129-L136), the context is `null`, which I suspect is due to the filters immediate closure of the scope and removal of the context from as a request property.

Am I on the right track here, when I suspect an extension of Quarkus' `OpenTelemetryClientFilter` is necessary? Or am I missing an obvious, maybe even existing, way to add client headers to the span of the REST client calls?

[a comment by an automated account is not shown]

**Reply by geoand** [reply]
cc @brunobat

**Comment by brunobat** [comment, marked as answer]
That is a feature that was never implemented of the REST client implementations. @svifred can you please create an issue with from this discussion?

Th implementation should be straight forward. 
We need to create the new properties to control the feature and add attributes to the span in this class:
https://github.com/quarkusio/quarkus/blob/6ccff484bc930f393a97f61de3391a8b1967ece6/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/intrumentation/restclient/OpenTelemetryClientFilter.java#L44
and this one as well:
https://github.com/quarkusio/quarkus/blob/e559cbdf93faac811d028f5af6acd5ad44b2f412/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/intrumentation/resteasy/OpenTelemetryClassicServerFilter.java#L21

**Reply by svifred** [reply]
@brunobat Here we go: [Improve OpenTelemetry instrumentation to support adding REST client headers to span for REST call](https://github.com/quarkusio/quarkus/issues/44307).
